# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data= pd.read_csv('/datasets/data.csv')
except:
    data= pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed']= data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children']!= -1) & (data['children']!= 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type']== t) & (data['days_employed'].isna()), 'days_employed']= \
    data.loc[(data['income_type']== t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income']= data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education']= data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data= data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(total_income):
    try:
        if 0<= total_income <= 30000:
            return 'E'
        elif 30001<= total_income <= 50000:
            return 'D'
        elif 50001<= total_income <= 200000:
            return 'C'
        elif 200001<= total_income <= 1000000:
            return 'B'
        elif total_income>= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category']= data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_grouped_children= data.groupby('children').agg({'debt': ['count', 'sum']})
data_grouped_children['conversion']= data_grouped_children['debt']['sum']/data_grouped_children['debt']['count']*100
data_grouped_children # Ваш код будет здесь. Вы можете создавать новые ячейки.

debt       conversion
          count   sum           
children                        
0         14091  1063   7.543822
1          4808   444   9.234609
2          2052   194   9.454191
3           330    27   8.181818
4            41     4   9.756098
5             9     0   0.000000

In [23]:
data_grouped_children= data.groupby('children').agg({'debt': ['count', 'sum']})
data_grouped_children['conversion']= data.groupby('children')['debt'].mean()*100
data_grouped_children

debt       conversion
          count   sum           
children                        
0         14091  1063   7.543822
1          4808   444   9.234609
2          2052   194   9.454191
3           330    27   8.181818
4            41     4   9.756098
5             9     0   0.000000

**Вывод:** 
1. Бездетные Клиенты кредиты отдают лучше, в этой группе самый низкий процент просрочки 7,54%.
2. В семьях с детьми вероятность возврата кредита, практически, на одном уровне, в среднем 9,15%. Поэтому, можно сказать, что семьи с детьми менее надежные заемщики. В таблице есть данные по семьям с 5-ю детьми, но выборка по ним не репрезентативна, поэтому данными можно пренебречь . Судя по тому, что наблюдается четкая зависимость между долей просрочки и наличием детей, 5 детей не станут исключением. 
3. Данные по семьям с детьми немного разнятся, поэтому хочется посмотреть ещё на зависимость от занятости замщиков.

In [24]:
data_grouped_income_type= data.groupby('income_type').agg({'debt': ['count', 'sum']})
data_grouped_income_type['conversion']= data_grouped_income_type['debt']['sum']/ data_grouped_income_type['debt']['count']*100
data_grouped_income_type

debt        conversion
                 count   sum            
income_type                             
безработный          2     1   50.000000
в декрете            1     1  100.000000
госслужащий       1451    86    5.926947
компаньон         5047   374    7.410343
пенсионер         3812   216    5.666317
предприниматель      2     0    0.000000
сотрудник        11015  1054    9.568770
студент              1     0    0.000000

**Вывод** 
1. Есть категории, в которых выборка не репрезентативна, и в целом не отражает реальной картины:
- безработный
- в декрете
- предприниматель
- студент
2. Среди остальных, сравнимых показателей, можно сказать, сотрудники, которые составляют 50% от всей выборки замщиков, имеют самый высокий процент невозвратов или просрочки кредита.
3. В категорию надежных заемщиков попали песионеры и госслужащие, доля невозвратов 5,66% и 5,92% соответственно. Хотя доходы их невысоки, но зато стабильные.
4. Компаньоны заняли среднюю позицию, доля 7,41%. Можно сказать, что кредиты возвращают на уровне бездетных семей.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
data_grouped_family_status= data.groupby('family_status').agg({'debt': ['count', 'sum']})
data_grouped_family_status['conversion']= data_grouped_family_status['debt']['sum']/ data_grouped_family_status['debt']['count']*100
data_grouped_family_status # Ваш код будет здесь. Вы можете создавать новые ячейки.

debt      conversion
                       count  sum           
family_status                               
Не женат / не замужем   2796  273   9.763948
в разводе               1189   84   7.064760
вдовец / вдова           951   63   6.624606
гражданский брак        4134  385   9.313014
женат / замужем        12261  927   7.560558

**Вывод:** 
1. Платежеспособность кредиторов зависит от семейного положения. 
- Семейные хорошо отдают кредиты, доля просрочки всего 7,56%.
- А вот самые низкие показатели у категорий "не женат / не замужем" и "гражданский брак", 9,76% и 9,31% соответственно. 

*Хотя выше, мы сделали вывод о том, что бездетные семьи более дисциплинированы в выплате кредитов. Значит, что бездетные и в браке - самые надежные заемщики.*  
2. Высокие показатели в категориях "вдовец/вдова" 6,62% и "в разводе"7,06%. Получается, что даже при отсутствии поддержки второй половины эти категории заемщиков более надежные. 


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
data_grouped_total_income= data.groupby('total_income_category').agg({'debt': ['count', 'sum']})
data_grouped_total_income['conversion']= data_grouped_total_income['debt']['sum']/ data_grouped_total_income['debt']['count']*100
data_grouped_total_income  # Ваш код будет здесь. Вы можете создавать новые ячейки

debt       conversion
                       count   sum           
total_income_category                        
A                         25     2   8.000000
B                       5014   354   7.060231
C                      15921  1353   8.498210
D                        349    21   6.017192
E                         22     2   9.090909

Исходя из диапазонов доходов, определили категории:
- 30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

**Выводы:**
1. Кредитополучатели от 30000-50000 (категория "D") самые надежные, доля просрочек 6,02%. Показатель заметно выбивается из всех остальных, новыборка по ним мала.
2. Самый высокий процент невозвратов в срок, у заемщиков с доходом менее 30000 руб. (категория "Е"), 9,09%. Но выборка по ним очень мала и сравнима с заемщиками с самым высоким уровнем дохода, свыше 1 млн.руб.(категория "А"), где доля просрочки 8%.
3. Очень большую долю, 74,5%, составляют заемщики с доходом 50000-200000 руб.(категория "С"), а доля просрочек 8,5%, то есть, практически, на уровне самых ненадежных заемщиков.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [27]:
data_grouped_purpose= data.groupby('purpose_category').agg({'debt': ['count', 'sum']})
data_grouped_purpose['conversion']= data_grouped_purpose['debt']['sum']/ data_grouped_purpose['debt']['count']*100
data_grouped_purpose  # Ваш код будет здесь. Вы можете создавать новые ячейки.

debt      conversion
                          count  sum           
purpose_category                               
операции с автомобилем     4279  400   9.347978
операции с недвижимостью  10751  780   7.255139
получение образования      3988  369   9.252758
проведение свадьбы         2313  183   7.911803

**Вывод:** 
1. Показатель возвращаемости кредитов на проведение свадьбы 7,9% и приобретение жилья 7,25% намного лучше, чем на остальные цели. Кстати, среди всех заемщиков, те, кто взял кредит под операции с недвижимостью составляют почти 50%, и они же самые надежные.  
2. Клиенты, взявшие кредит на покупку автомобиля или получение образования отдают кредиты хуже, доля просрочек 9,35% и 9,25% соответственно.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

В исходных данных в двух столбцах: `days_employed` -общий трудовой стаж в днях и `total_income`  - ежемесячный доход, есть пропущенные значения, и они одинаковы, 2174, что составляет 10% от всех 21525 строк с данными. Допустимое значение,около 1%. Между двумя этими показателями есть прямая зависимость: нет стажа, нет дохода.

Пропуски имеются в категориях занятости: "Сотрудник", "Пенсионер", "Компаньон" и "Госслужащий".

Возможные причины появления пропусков в данных:
1. Все эти заемщики, как правило, с невысоким доходом, и речь идет о небольших суммах кредита, получить которые можно без справок и доходах.
2. У каждого банка есть продукты (кредитные карты, например), с помощью которых можно получить ссуду без официального трудоустройства, но на определенных условиях.
Именно потому, что такие данные не обязательны в указанных случаях, поэтому они отсутствуют в выборке.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
- Медианное значение наиболее точно отражает статистику в случае, когда в вариативном ряду есть отклонения от средних величин в большую и меньшую стороны. Медианное значение дает представление о типичном значении в определенной выборке.
В нашем исследовании мы применяли медиану к значениям:
- days_employed — общий трудовой стаж в днях и 
- total_income — ежемесячный доход.
Понятно, что именно эти две величины содержат данные, в которых большой разброс. Стаж может быть, как 3 месяца, так и 30 лет. Заработная плата от 0 до 1 млн. или более. Поэтому брать среднюю величину для расчета данных показателей не верно, оно сильно тсказит реальную картину. Именно медиана даст представление о самой типичной величине в каждой из выборок.


### Шаг 4: общий вывод.

Нами было проведено исследование, заказчиком которого выступил кредитный отдел банка, с целью выявить влияние различных факторов на факт погашения кредита в срок. 

Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

В начале исследования было выдвинуто несколько гипотез:
- Гипотеза 1: Количество детей влияет на возврат кредита в срок
- Гипотеза 2: Семейное положение влияет на возврат кредита в срок
- Гипотеза 3: Уровень дохода влияет на возврат кредита в срок
- Гипотеза 4: Цель кредита влияет на возврат кредита в срок

Проведя исследования по предоставленным от банка данным, выявили следующее:

**Гипотеза 1 частично подтверждена:** Выборка по клиентам с 3-мя, 4-мя и 5-ю детьми мала, поэтому данных нужно больше.
 
   - 7,54% - доля просроченных кредитов у бездетных заёмщиков
   - 9,23% - доля просроченных кредитов у Клиентов с 1-им ребенком
   - 9,45% - доля просроченных кредитов у Клиентов с 2-мя детьми

*Рекомендация:* Провести дополнительные исследования на большей выборке данных.

**Гипотеза 2 подтверждена:** 
   
   - 9,76% и 9,31%  - доли соответственно в категориях "не женат / не замужем" и "гражданский брак", являются более рисковыми заемщиками
   - 7,56% - доля просрочки у семейных клиентов, "вдовец/вдова" 6,62% и "в разводе"7,06%. В этих группах Клиентов риски стать должниками ниже

*Рекомендация:* Доля клиентов, не состоящих в браке, состаявляет почти 50% от заёмщиков, находящихся в законном браке. Кроме того, разница между 2-мя выводами невелика, 3%, поэтому не стоит придавать весомое значение этой зависимости при принятии решения о кредитоспособности.

**Гипотеза 3 частично подтверждена:** Требуется больше данных по клиентам категорий A(более 1000000), D(30000-50000), E(0-30000) для более точных выводов.
   
   - 8,5% - доля просроченных кредитов у заемщиков с доходом 50000-200000 руб.(категория "С")
   - 7,06% - доля просроченных кредитов у заеьщиков с доходом 200001-1000000 руб.(категория В).

*Рекомендация:* 
- Для более точных выводов необходимо провести дополнительное исследование на бОльшей выборке. 
- Кроме того, предлагаем разбить категорию "С", и выделить в ней 2 группы: 50000-100000 и 100001-200000, так как средняя зарплата по России по данным Росстата за 2021 составляет 57 244 руб. Вторая группа будет "очищена" от среднестатиститческих данных, и показатели статнут более реалистичными.

**Гипотеза 4 подтверждена:** 

   - 9,35% - доля просроченных кредитов под покупку автомобиля 
   - 7,26% - доля просроченных кредитов под операции с недвижимостью
   - 9,25% - доля просроченных кредитов под получение образования
   - 7,91% - доля просроченных кредитов под проведение свадьбы 

*Рекомендация:* Клиенты, взявшие кредит на покупку автомобиля или получение образования отдают кредиты хуже, но при этом, они составляют больше половины (8267) от доли заемщиков (14739) с лучшим показателем возврата кредита, и разница в процентном соотношении между этими группами невелика, 2%. Поэтому не стоит уделять весомое значение этой зависимости при принятии решения о кредитоспособности.

**ОБЩИЕ РЕКОМЕНДАЦИИ**
1. Специалистам кредитного отдела при построении модели кредитного скоринга можно использовать выводы, полученные из 2-ой и 4-ой гипотез: зависмость от семейного положения и целей кредита.
2. Для проверки 1-ой и 3-ей гипотез, зависимость от количества детей и дохода заемщика, провести дополнительные исследования на большем количестве данных.
*Из дополнительного исследования: наблюдается зависимость возвратов кредита от занятости Клиента. Обратить внимание на категорию "Сотрудники", которые составляют 50% от всей выборки заемщиков, но при этом имеют самый высокий процент невозвратов 9,57%. В качестве рекомендации - провести дополнительное исследвание надежности от стажа работы.*
